pkg/api/handlers/mcp.go

```diff
@@ -742,15 +716,137 @@ func (m *MCPHandler) GetPrompt(req api.Context) error {
 	})
 }
 
-func ServerFromMCPServerInstance(req api.Context, instanceID string) (v1.MCPServer, mcp.ServerConfig, error) {
-	var (
-		server   v1.MCPServer
-		instance v1.MCPServerInstance
-	)
-	if err := req.Get(&instance, instanceID); err != nil {
-		return server, mcp.ServerConfig{}, err
+func ServerForActionWithConnectID(req api.Context, id string) (string, v1.MCPServer, mcp.ServerConfig, error) {
+	switch {
+	case system.IsMCPServerInstanceID(id):
+		server, config, err := serverFromMCPServerInstanceID(req, id)
+		return id, server, config, err
+	case system.IsMCPServerID(id):
+		var server v1.MCPServer
+		if err := req.Get(&server, id); err != nil {
+			return "", server, mcp.ServerConfig{}, err
+		}
+
+		if server.Spec.MCPCatalogID != "" || server.Spec.PowerUserWorkspaceID != "" {
+			// This is a multi-user MCP server, and user is trying to connect to it.
+			// List the MCP server instances, sort by creation time, and take the first one.
+			var instances v1.MCPServerInstanceList
+			if err := req.List(&instances, &kclient.ListOptions{
+				FieldSelector: fields.SelectorFromSet(map[string]string{
+					"spec.mcpServerName": id,
+					"spec.userID":        req.User.GetUID(),
+				}),
+			}); err != nil {
+				return "", server, mcp.ServerConfig{}, err
+			}
+			if len(instances.Items) == 0 {
```

Comment:
is it possible for multiple users to trigger this at the same time?

Reply:
Yes, but this is scoped to one user by the field selector. Even if this is triggered multiple times simultaneously by the same user, the result would change. The URL would refer to the first created server.